untyped-lambda-calculus/article.md

```diff
@@ -0,0 +1,493 @@
+
+> ... one of the myths concerning LISP that people think up or invent for themselves becomes apparent, and that is that LISP is somehow a realization of the lambda calculus, or that was the intention.
+> The truth is that I didn't understand the lambda calculus, really.
+>
+> -- John McCarthy, Lisp session, History of Programming Languages
+
+Also, while not strictly speaking based on lambda calculus, C++ template metalanguage is a functional language in its own right, so much of what we discuss below happens to be applicable.
+In fact, one could either embed lambda calculus into C++ template metalanguage or write a transpiler from a lambda calculus-based language into C++ template metalanguage.
+See, for instance, [Translating Lambda Calculus into C++ Templates by Vít Šefl](https://dx.doi.org/10.1007/978-3-030-83978-9_5).
+
+## A bit of history
+
+> 1936 - Alan Turing invents every programming language that will ever be but is shanghaied by British Intelligence to be 007 before he can patent them.
+>
+> 1936 - Alonzo Church also invents every language that will ever be but does it better.
+> His lambda calculus is ignored because it is insufficiently C-like.
+> This criticism occurs in spite of the fact that C has not yet been invented.
+>
+> -- James Iry, [A Brief, Incomplete, and Mostly Wrong History of Programming Languages][brief-incomplete-history]
+
+[brief-incomplete-history]: http://james-iry.blogspot.com/2009/05/brief-incomplete-and-mostly-wrong.html
+
+Lambda calculus, initially envisioned as a formal logic system, was developed by Alonzo Church around the 1930s to explore the foundations of mathematics.
+The initial formulation had a logical inconsistency known as [Kleene–Rosser paradox] (for a more comprehensive overview, refer to, e.g. [Paradoxes and Contemporary Logic][paradoxes-contemporary-logic]).
+To sidestep this issue, Church isolated the part of lambda calculus relevant only to computation in 1936.
+This isolate is now known as the untyped lambda calculus.
+Later, In 1940, a typed version of lambda calculus, based on Russel's type theory, was introduced, which has weaker expressive power but is logically consistent.
+
+[Kleene–Rosser paradox]: https://en.wikipedia.org/wiki/Kleene%E2%80%93Rosser_paradox
+[paradoxes-contemporary-logic]: https://plato.stanford.edu/entries/paradoxes-contemporary-logic/#IncoCertFormLogi
+
+The lambda symbol used in the standard notation, from which lambda calculus derives its name, most likely was chosen arbitrarily.
+However, there's an apocryphal story the notation was derived from $\hat x$, morphing first into $\Lambda x$, and later into $\lambda x$.
+
+In the mid-1960s, Peter Landin has shown that lambda calculus models arbitrarily complex programming languages.
+This insight arguably kickstarted the research on functional programming languages.
+
+## The pure untyped lambda calculus
+
+The untyped lambda calculus is, first and foremost, a consistent expression rewriting system.
```

Comment:
Is "consistent system for rewriting expressions" correct? I feel it would sound better.

Reply:
Technically, it's the same words in slightly different order, so it's not wrong. Can't say I like it, but that's why we have an editor, I guess :laughing:

Fixed.